language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/createvar/CreateVariableWithTypeCodeAction.java

```diff
@@ -0,0 +1,283 @@
+import io.ballerina.compiler.syntax.tree.ClientResourceAccessActionNode;
+import io.ballerina.compiler.syntax.tree.NamedArgumentNode;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.RemoteMethodCallActionNode;
+import io.ballerina.compiler.syntax.tree.SimpleNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.projects.util.ProjectConstants;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.codeaction.CodeActionNodeValidator;
+import org.ballerinalang.langserver.codeaction.CodeActionUtil;
+import org.ballerinalang.langserver.common.ImportsAcceptor;
+import org.ballerinalang.langserver.common.constants.CommandConstants;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.common.utils.FunctionGenerator;
+import org.ballerinalang.langserver.common.utils.NameUtil;
+import org.ballerinalang.langserver.common.utils.PositionUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagBasedPositionDetails;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionKind;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Create variable code action when type infer diagnostic is presented.
+ *
+ * @since 2201.4.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class CreateVariableWithTypeCodeAction extends CreateVariableCodeAction {
+
+    private static final String NAME = "Create variable with type";
+
+    @Override
+    public boolean validate(Diagnostic diagnostic, DiagBasedPositionDetails positionDetails,
+                            CodeActionContext context) {
+        return diagnostic.diagnosticInfo().code().equals("BCE3934")
+                && context.currentSemanticModel().isPresent()
+                && context.nodeAtRange().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
+                || context.nodeAtRange().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
+                || context.nodeAtRange().parent().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
+                || context.nodeAtRange().parent().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
+                || context.nodeAtRange().parent().parent().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
+                || context.nodeAtRange().parent().parent().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
+                && CodeActionNodeValidator.validate(context.nodeAtRange());
```

Comment:
Shall we refactor this code? We can move
```
context.nodeAtRange().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
                || context.nodeAtRange().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
                || context.nodeAtRange().parent().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
                || context.nodeAtRange().parent().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
                || context.nodeAtRange().parent().parent().kind() == SyntaxKind.CLIENT_RESOURCE_ACCESS_ACTION
                || context.nodeAtRange().parent().parent().kind() == SyntaxKind.REMOTE_METHOD_CALL_ACTION
```
part into a separate method called `isInRemoteMethodCallOrResourceAccess()` and inside that we can iterate through the parents starting from the `nodeAtRange()` until we a remote method call action or client resource access action.